File: cliente.py
class Cliente:
    def __init__(self, _dni, _nombre) -> None:
        self.dni = _dni
        self.nombre = _nombre

    def __eq__(self, value: object) -> bool:
        if type(self) != type(value):
            return NotImplemented
        return self.dni == value.dni

    def __hash__(self) -> int:
        return hash(self.dni)


c1 = Cliente('123', 'Pepe')
c2 = Cliente('123', 'María')

print(hash(c1)) # print(c1.__hash__())
print(hash(c2))

print(c1 == c2) # print(c1.__eq__(c2))

File: java/banco/clientes/Cliente.java
package clientes;

import java.util.Objects;

public class Cliente implements Cloneable {
    private String dni;
    private String nombre;
    private String apellidos;

    public Cliente(String dni, String nombre, String apellidos) {
        this.dni = dni;
        this.nombre = nombre;
        this.apellidos = apellidos;
    }

    @Override
    protected Cliente clone() throws CloneNotSupportedException {
        return (Cliente) super.clone();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }

        // if (obj == null) {
        //     return false;
        // }

        if (!(obj instanceof Cliente)) {
            return false;
        }

        // if (getClass() != obj.getClass()) {
        //     return false;
        // }

        Cliente c = (Cliente) obj;

        return Objects.equals(dni, c.dni);
    }

    @Override
    public int hashCode() {
        return Objects.hash(dni);
    }

    public String getDni() {
        return dni;
    }

    public void setDni(String dni) {
        this.dni = dni;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getApellidos() {
        return apellidos;
    }

    public void setApellidos(String apellidos) {
        this.apellidos = apellidos;
    }
}
